### PR TITLE
Add Sender.queue_size limits

### PR DESCRIPTION
**Pull request by djmitche**
This adds a queue_size limit to the Sender class, with messages beyond that limit dropped.  We support Ruby-2.1.0, which doesn't have a SizedQueue that supports "attempted" sends (it always blocks the sender if the queue is full), so we need to limit the queue size out-of-band.  Also, we want to limit the size based on bytes, not message count.  Because the `@message_queue_bytesize` is accessed concurrently in a read-modify-write pattern, it requires a mutex.

Things I'd like feedback on:
 * Ideas on other ways to do this that don't involve two mutex syncs on each message
 * The tests - is there something similar to Waiter already defined?  Or another more rspec-y way to test this kind of concurrency-related behavior?
 * Logging on drops - this will log every drop, which is probably exactly the wrong thing to do.  Other options:
   * Log only the first drop
   * Log edges (not dropping -> dropping / dropping -> not dropping)
   * Log the first and then every N messages dropped
   * <your idea here>

This isn't quite done yet.  Still to do:
 * [x] Thread queue_size through Statsd constructor
 * [x] More tests, at least covering resuming sending
 * [x] Docs
 * [x] Changelog

### CODE REVIEW

**Comment by remeh** [comment]
Small feedback @djmitche 

* I don't think Ruby is supporting any sort of atomic integer whatsoever so I would probably have done the same...
* Can't help here sorry
* I think we want to use the internal telemetry of the client to report the amount of drops. However about the logs, that's a good question, we should check what's the behavior in our other implementations (i.e. datadog-go). 

You have been with the byte size and not the messages count, I would probably have done the same, but it may make sense to validate it has been done the same way in other implementations and to discuss about it with the team otherwise.

**Comment by djmitche** [comment]
datadog-go doesn't log anything, but does count it in `datadog.dogstatsd.client.packets_dropped_queue`.  I'll do something similar in this client.

**Comment by djmitche** [comment]
This is based on the Go client, which uses [`WithSenderQueueSize`](https://pkg.go.dev/github.com/DataDog/datadog-go/v5@v5.0.2/statsd#WithSenderQueueSize):

> WithSenderQueueSize sets the size of the sender queue in number of buffers.
> 
> After data has been serialized in a buffer they're pushed to a queue that the sender will consume and then each one ot the agent.
> 
> The default value 0 will set the option to the optimal size for the transport protocol used: 2048 for UDP and named pipe and 512 for UDS.

So that should be easy to emulate with the ruby version.

**Comment by djmitche** [comment]
Also, somehow I convinced myself that Ruby doesn't have a GIL.  It does.  So access to an integer is automatically serialized by that GIL.

**Comment by djmitche** [comment]
I've asked for some advice on how best to test this.  I don't think Waiter is a good idea.

**Comment by djmitche** [comment]
Ivo has suggested using dependency-injection to inject a fake Queue implementation here, mocking out the background thread's behavior.  I had tried mocking Thread, without success, so I'll try Queue instead.

**Comment by djmitche** [comment]
Supporting those ancient Ruby versions is a real drag.

**Comment by djmitche** [comment]
Hm, it seems webrick might have dropped support for rub-2.0??

**Comment by djmitche** [comment]
I think the Ruby-2.0 failures are unrelated to this, so maybe we can handle that with #229 ?

**Comment by djmitche** [comment]
Merged from main -- looks like I changed all the same places as @abicky :)

File: spec/statsd/telemetry_spec.rb [hunk 1]
```diff
@@ -98,10 +103,10 @@
         skip 'Ruby too old' if RUBY_VERSION < '2.3.0'
       end
 
-      it 'makes only 8 allocations' do
+      it 'makes only 10 allocations' do
```
Comment (remeh) [inline spec/statsd/telemetry_spec.rb:106]: nit: `12`

**Review by remeh — APPROVED** [review]
Nice!

**Review by remeh — APPROVED** [review]

**Pull request merged**